src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs

```diff
@@ -0,0 +1,192 @@
+            LogLevel.Information,
+            EventIds.PipelineEnd,
+            "End processing HTTP request after {ElapsedMilliseconds}ms - {StatusCode}");
+
+        private static bool GetDisableUriRedactionSettingValue()
+        {
+            if (AppContext.TryGetSwitch("System.Net.Http.DisableUriRedaction", out bool value))
+            {
+                return value;
+            }
+
+            string? envVar = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_NET_HTTP_DISABLEURIREDACTION");
+
+            if (bool.TryParse(envVar, out value))
+            {
+                return value;
+            }
+            else if (uint.TryParse(envVar, out uint intVal))
+            {
+                return intVal != 0;
+            }
+
+            return false;
+        }
+
+        public static void LogRequestStart(this ILogger logger, HttpRequestMessage request, Func<string, bool> shouldRedactHeaderValue)
+        {
+            // We check here to avoid allocating in the GetUriString call unnecessarily
+            if (logger.IsEnabled(LogLevel.Information))
+            {
+                _requestStart(logger, request.Method, GetRedactedUriString(request.RequestUri), null);
+            }
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.RequestHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Request, request.Headers, request.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static void LogRequestEnd(this ILogger logger, HttpResponseMessage response, TimeSpan duration, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestEnd(logger, duration.TotalMilliseconds, (int)response.StatusCode, null);
+
+            if (logger.IsEnabled(LogLevel.Trace))
+            {
+                logger.Log(
+                    LogLevel.Trace,
+                    EventIds.ResponseHeader,
+                    new HttpHeadersLogValue(HttpHeadersLogValue.Kind.Response, response.Headers, response.Content?.Headers, shouldRedactHeaderValue),
+                    null,
+                    (state, ex) => state.ToString());
+            }
+        }
+
+        public static IDisposable? BeginRequestPipelineScope(this ILogger logger, HttpRequestMessage request)
+        {
+            return _beginRequestPipelineScope(logger, request.Method, GetRedactedUriString(request.RequestUri));
+        }
+
+        public static void LogRequestPipelineStart(this ILogger logger, HttpRequestMessage request, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineStart(logger, request.Method, GetRedactedUriString(request.RequestUri), null);
```

Comment:
We were using `AbsoluteUri` before, which would get cached by the `Uri`.
With `GetRedactedUriString` that's not the case, and we might be allocating it three times. Can we avoid that here?

Reply:
I was able to save one of the `LoggingScopeHttpMessageHandler` allocations in [4b1d04](https://github.com/dotnet/runtime/pull/103769/commits/74b1d040d4a48cf08c201d4936f80d766aa1b9c0).

Unfortunately, there is no sane way that would allow us to flow the uri string further to `LoggingHttpMessageHandler`.

Reply:
It's unfortunate, but makes sense. Thanks.